Add SVE128 ISA

Nicoshev opened the pull request:
Summary: Partly Importing and adapting https://github.com/pytorch/pytorch/pull/138388, adding SVE128 as ISA.

Intention is to add SVE128 translation layers for Vectorized data types.
Idea is to have 1 PR per file, aside from the current one, plus a last one modifying cmake files to enable the new ISA selectively.

Tested current changes on a nightly run, to verify no regressions occur on systems leveraging SVE256.

No regressions spotted when running test_ops.py, a set of 34k unit tests. A machine leveraging SVE128 was used towards this testing.



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

fadara01 reviewed (commented):
It's great the you guys are looking into this. We do see a good perf uplift for SVE128 vs NEON when we compare both on Neonverse-V2.

aditew01 commented:
@Nicoshev quick question on the scope for the PR. 
Are we in process of testing if this will be a good idea or do we want to enable this and push to main?

aditew01 commented:
> Hey @aditew01, intention is to add SVE128 translation layers for Vectorized data types You can see them all in this PR: https://github.com/pytorch/pytorch/pull/160329/files

thanks for sharing the plan. sounds good to me!

Ryo-not-rio commented on aten/src/ATen/Version.cpp:
```diff
@@ -103,9 +103,17 @@ std::string get_cpu_capability() {
+      return "SVE128";
+    case native::CPUCapability::SVE256:
+      return "SVE256";
 #elif defined(HAVE_SVE256_CPU_DEFINITION) && defined(HAVE_ARM_BF16_CPU_DEFINITION)
     case native::CPUCapability::SVE256:
       return "SVE256";
+#elif defined(HAVE_SVE128_CPU_DEFINITION) && defined(HAVE_ARM_BF16_CPU_DEFINITION)
+    case native::CPUCapability::SVE128:
+      return "SVE128";
```
We should plan to get rid of `HAVE_SVE128_CPU_DEFINITION` and `HAVE_SVE256_CPU_DEFINITION` and just use `HAVE_SVE_CPU_DEFINITION` since the other two will always be true when compiling on any sve machines

Ryo-not-rio reviewed (commented):
Thank you for adding SVE128 support, my only note would be that we should be unifying all `HAVE_SVE{vec_len}_CPU_DEFINITION` to just `HAVE_SVE_CPU_DEFINITION` and also to enable this flag only if BF16 support is available. This would make the whole codegen logic much cleaner

malfet commented on torch/_inductor/cpu_vec_isa.py:
```diff
@@ -168,9 +168,29 @@ def __str__(self) -> str:
+    _dtype_nelements = {torch.float: 4, torch.bfloat16: 8, torch.float16: 8}
+
+    def __str__(self) -> str:
+        if config.is_fbcode():
```
Not that it matter much, but why?

malfet commented on torch/_inductor/cpu_vec_isa.py:
```diff
@@ -156,7 +156,7 @@ def __bool__impl(self, vec_isa_ok) -> bool:
 class VecNEON(VecISA):
     _bit_width = 128  # This is required to leverage the compute implemented in aten/src/ATen/cpu/vec/vec128/vec128_float_neon.h
-    _macro = ["CPU_CAPABILITY_NEON", "AT_BUILD_ARM_VEC256_WITH_SLEEF"]
+    _macro = ["CPU_CAPABILITY_NEON", "AT_BUILD_ARM_VECSVE_WITH_SLEEF"]
```
Hmm, why is this part of NEON ISA?

Nicoshev replied:
@malfet It is just an abuse on the flag. NEON Sleef routines are used on the Neon ISA, it is just that the same flag is checked for both.

malfet commented on aten/src/ATen/native/cpu/BlasKernel.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <ATen/native/CPUBlas.h>
 #include <ATen/native/cpu/zmath.h>
 #include <ATen/native/cpu/ReducedPrecisionFloatGemvFastPathKernel.h>
+#include <ATen/native/DispatchStub.h>
```
Why this change is needed?
```suggestion
```

malfet commented on aten/src/ATen/native/cpu/AmpGradScalerKernels.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <ATen/Dispatch.h>
 #include <ATen/OpMathType.h>
 #include <ATen/core/Tensor.h>
+#include <ATen/native/DispatchStub.h>
```
Why this change is needed?
```suggestion
```

malfet commented on aten/src/ATen/native/DispatchStub.cpp:
```diff
@@ -39,19 +39,20 @@ static CPUCapability compute_cpu_capability() {
     int sve_vl = cpuinfo_get_max_arm_sve_length(); //Returns maximum SVE VL supported by your HW.
-#ifdef HAVE_SVE256_CPU_DEFINITION
-    if (envar == "sve256") {
+    if (envar == "sve" || envar == "sve256" || envar == "sve128") {
```
Hmm, can you elaborate what you are trying to do here? It feels like this change would not allow me to set ISA to SEV128 if my CPU support SVE256, which is wrong, isn't it?
Also, why introduce `sve` option? Auto-detect is indeed nice, but for this to work one simply should not define any environment variable at all

Nicoshev replied:
Idea is to ultimately just have "sve", but the other ones were left to avoid having to introduce bigger changes.

If your CPU has a 256-bit reg length, it should only use SVE256

malfet commented on aten/src/ATen/cpu/vec/vec128/vec128_convert.h:
```diff
@@ -4,7 +4,8 @@
 namespace at::vec {
 inline namespace CPU_CAPABILITY {
-#if (defined(__aarch64__) && !defined(CPU_CAPABILITY_SVE256))
+#if defined(__aarch64__) && \
```
Is it identical to what it was before? But just unnecessarily more comples?

Nicoshev commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -78,17 +88,17 @@ Windows llvm will not have this definition.
 #if defined(__GNUC__)
-#define __at_align__ __attribute__((aligned(32)))
+#define __at_align__ __attribute__((aligned(16)))
 #elif defined(_WIN32)
-#define __at_align__ __declspec(align(32))
+#define __at_align__ __declspec(align(16))
```
Because this case is now #define VECTOR_WIDTH 16

Nicoshev commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -194,7 +204,7 @@ struct Vectorized {
       typename... Args,
       typename = std::enable_if_t<(sizeof...(Args) == size())>>
-  Vectorized(Args... vals) : values{vals...} {}
+  Vectorized(Args... vals) : values{static_cast<T>(vals)...} {}
```
In case there is a constructor taking elements of a different type and performing a cast in another way

Nicoshev committed:
[PyTorch] Add SVE128 ISA (#158932)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/158932

Importing https://github.com/pytorch/pytorch/pull/138388, as it improves SVE support for perfkernels

Test Plan: We will test it on AdFinder/AdRetriever/AdRanker offline tier

Reviewed By: r1mikey

Differential Revision: D70788867
